File: wishlistservice/src/main/java/com/cgi/wishlistservice/service/WishlistService.java
package com.cgi.wishlistservice.service;

import com.cgi.wishlistservice.exception.WishlistItemNotFoundException;
import com.cgi.wishlistservice.exception.WishlistServiceException;
import com.cgi.wishlistservice.model.WishlistItem;
import com.cgi.wishlistservice.repository.WishlistRepository;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class WishlistService {

    @Autowired
    private WishlistRepository wishlistRepository;

    public WishlistItem addToWishlist(WishlistItem wishlistItem) {
    	 try {
             return wishlistRepository.save(wishlistItem);
         } catch (Exception e) {
             throw new WishlistServiceException("Failed to add item to wishlist", e);
         }
    }
    public List<WishlistItem> getWishlistByTrackName(String trackName) {
    	  try {
              return wishlistRepository.findByTrackName(trackName);
          } catch (Exception e) {
              throw new WishlistServiceException("Failed to fetch wishlist by track name", e);
          } // Fetch items by song name
    }
    public List<WishlistItem> getUserWishlist(String userId) {
    	  try {
              return wishlistRepository.findByUserId(userId);
          } catch (Exception e) {
              throw new WishlistServiceException("Failed to fetch wishlist for user", e);
          }
    }

    public void removeFromWishlist(String itemId) {
    	 Optional<WishlistItem> item = wishlistRepository.findById(itemId);
         if (!item.isPresent()) {
             throw new WishlistItemNotFoundException("Item with ID " + itemId + " not found in wishlist");
         }
         try {
             wishlistRepository.deleteById(itemId);
         } catch (Exception e) {
             throw new WishlistServiceException("Failed to remove item from wishlist", e);
         }  // Ensure itemId is String
    }
}

File: frontend-version-4-master/src/app/wishlist.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class WishlistService {

  private wishlistUrl: string = 'http://localhost:8084/wishlist';// URL for the Wishlist Microservice


  constructor(private http: HttpClient) { }

  // Method to add a song to the wishlist
  addToWishlist(wishlistItem: any): Observable<any> {
    return this.http.post(`${this.wishlistUrl}/add`, wishlistItem);
  }
  // Function to fetch wishlist items for a given user ID
  // Get the wishlist for a specific song by track name
  getWishlistByTrackName(trackName: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.wishlistUrl}/song/${trackName}`);
  }
 // Method to get all wishlist items for a user
 getUserWishlist(userId: number): Observable<any[]> {
  return this.http.get<any[]>(`${this.wishlistUrl}/user/${userId}`);
}
  removeFromWishlist(trackId: string): Observable<any> {
    return this.http.delete(`${this.wishlistUrl}/remove/${trackId}`);
  }
}
